bins/stackabled/src/middleware/authentication.rs

```diff
@@ -0,0 +1,173 @@
+use std::{
+    borrow::Borrow,
+    collections::HashMap,
+    path::{Path, PathBuf},
+    sync::{Arc, RwLock},
+};
+
+use axum::{
+    body::BoxBody,
+    extract::Extension,
+    headers::{
+        authorization::{Basic, Bearer},
+        HeaderMapExt,
+    },
+    http::{HeaderMap, HeaderValue, StatusCode},
+    response::IntoResponse,
+    Json,
+};
+use snafu::{ResultExt, Snafu};
+use tower_http::validate_request::{ValidateRequest, ValidateRequestHeaderLayer};
+use uuid::Uuid;
+
+#[derive(PartialEq, Eq, Hash, Clone)]
+pub struct Username(String);
+enum PasswordHash {
+    Bcrypt(String),
+}
+#[derive(PartialEq, Eq, Hash, Clone)]
+struct SessionToken(String);
+
+impl Borrow<str> for Username {
+    fn borrow(&self) -> &str {
+        &self.0
+    }
+}
+impl Borrow<str> for SessionToken {
+    fn borrow(&self) -> &str {
+        &self.0
+    }
+}
+
+#[derive(Default)]
+struct State {
+    accounts: HashMap<Username, PasswordHash>,
+    sessions: RwLock<HashMap<SessionToken, Username>>,
+}
+
+#[derive(Default, Clone)]
+pub struct Authenticator {
+    state: Arc<State>,
+}
+
+#[derive(Debug, Snafu)]
+#[snafu(module)]
+pub enum HtpasswdError {
+    #[snafu(display("failed to read htpasswd file at {path:?}"))]
+    Read {
+        source: std::io::Error,
+        path: PathBuf,
+    },
+    #[snafu(display("malformed htpasswd entry on line {line}"))]
+    Entry {
+        source: HtpasswdEntryError,
+        line: usize,
+    },
+}
+#[derive(Debug, Snafu)]
+#[snafu(module)]
+pub enum HtpasswdEntryError {
+    #[snafu(display("invalid hash type (only bcrypt is currently supported)"))]
+    InvalidHashType,
+    #[snafu(display("no username/password separator"))]
+    NoSeparator,
+}
+
+impl Authenticator {
+    pub fn load_htpasswd(path: &Path) -> Result<Self, HtpasswdError> {
+        use htpasswd_entry_error::*;
+        use htpasswd_error::*;
+        let htaccess = std::fs::read_to_string(path).context(ReadSnafu { path })?;
+        let mut accounts = HashMap::new();
+        for (line, entry) in htaccess.lines().enumerate() {
+            if let Some((username, prefixed_pw_hash)) = entry.split_once(':') {
+                if prefixed_pw_hash.starts_with("$2y$") {
+                    accounts.insert(
+                        Username(username.to_string()),
+                        PasswordHash::Bcrypt(prefixed_pw_hash.to_string()),
+                    );
+                    Ok(())
+                } else {
+                    InvalidHashTypeSnafu.fail()
+                }
+            } else {
+                NoSeparatorSnafu.fail()
+            }
+            .context(EntrySnafu { line })?
```

Comment:
This implementation looks solid, but is there a crate out there to pass these files? If not, we can keeps it. It look simple enough.

Reply:
That sounds reasonable. It might be worth to pull this code out into its own file and maybe add some tests. In the future we might even put it in a separate crate.

Reply:
741c431842a533579073f134e66ab3dfa7b900e4